Gems/Archive/Code/Source/Clients/ArchiveReader.cpp

```diff
@@ -0,0 +1,885 @@
+            const size_t availableBytesInCompressedBlock = AZStd::min(compressedBlockRemainingSpan.size(),
+                ArchiveBlockSizeForCompression);
+            AZStd::span<const AZStd::byte> compressedDataForBlock = compressedBlockRemainingSpan
+                .first(availableBytesInCompressedBlock)
+                .first(blockCompressedSize);
+            // Slide the compressed block remaining span by 2 MiB
+            compressedBlockRemainingSpan.subspan(availableBytesInCompressedBlock);
+
+            // Get the block span for storing the decompressed block
+            // As the uncompressed size is 2 MiB for all blocks except the last
+            // the entire contiguous file sequence will be available in the decompressedResultSpan after the loop
+            const size_t remainingBytesInBlockSpan = AZStd::min(decompressionRemainingSpan.size(),
+                ArchiveBlockSizeForCompression);
+            AZStd::span<AZStd::byte> decompressionBlockSpan = decompressionRemainingSpan
+                .first(remainingBytesInBlockSpan);
+            // Slide the remaining decompressed span by 2 MiB as well
+            decompressionRemainingSpan.subspan(remainingBytesInBlockSpan);
+
+            const AZ::u32 decompressTaskCount = AZStd::min(static_cast<AZ::u32>(blockRange.second - blockRange.first),
+                maxDecompressTasks);
+
+            // Task graph event used to block decompressing blocks in parallel
+            auto taskDecompressGraphEvent = AZStd::make_unique<AZ::TaskGraphEvent>("Content File Decompress Sync");
+            AZ::TaskGraph taskGraph{ "Archive Decompress Tasks" };
+            AZ::TaskDescriptor decompressTaskDescriptor{ "Decompress Block", "Archive Content File Decompression" };
+            for (AZ::u32 decompressTaskSlot = 0; decompressTaskSlot < decompressTaskCount; ++decompressTaskSlot)
+            {
+                //! Decompress Task to execute in task executor
+                auto decompressTask = [
+                    decompressionInterface, &decompressionOptions, decompressionBlockSpan,
+                        compressedDataForBlock,
+                        &decompressedBlockResult = decompressedBlockResults[decompressTaskSlot]]()
+                {
+                    // Decompressed the compressed block
+                    decompressedBlockResult = decompressionInterface->DecompressBlock(
+                        decompressionBlockSpan, compressedDataForBlock, decompressionOptions);
+                };
+
+                    taskGraph.AddTask(decompressTaskDescriptor, AZStd::move(decompressTask));
+                    // Increment to the block index
+                    ++blockRange.first;
+            }
+
+            taskGraph.SubmitOnExecutor(m_taskExecutor, taskDecompressGraphEvent.get());
+            // Sync on the task completion
+            taskDecompressGraphEvent->Wait();
+
+            // Validate the decompression for all blocks
```

Comment:
Is it the responsibility of the decompressor to store a CRC of the decompressed contents alongside the data and verify that the decompressed contents match after decompression?

Reply:
I added support for writing a CRC32 for each file added to the Archive in this commit https://github.com/o3de/o3de/pull/16129/commits/efaf58563abac11f3eddd9548b6620a8b9593412